### PR TITLE
fixed the issue with the performance of mul! on matrix type QRCompact…

### PR DESCRIPTION
**Pull request by jmkl009**
…WYQ and a vector

### CODE REVIEW

**Comment by kshyatt** [comment]
#31147 - should be a fix for this! A HackIllinois PR :)

**Comment by jmkl009** [comment]
Performance comparison:

n, m = 500, 10
Q = qr(randn(n, m)).Q;
x = randn(n)
y = randn(n)
@time mul!(y, Q, x); #2.9s before 
   0.000036 seconds (5 allocations: 320 bytes)
@time y = Q*x; 
   0.000056 seconds (6 allocations: 4.375 KiB)

**Review by martinholters — COMMENTED** [review]
Huh, isn't this changing `B`?

**Comment by jmkl009** [comment]
> Huh, isn't this changing `B`?

Fixed, I originally thought it was okay to change B in a function with an exclamation mark.

File: stdlib/LinearAlgebra/src/qr.jl [hunk 1]
```diff
@@ -522,6 +522,10 @@ end
 
 ## Multiplication by Q
 ### QB
+QRCompactWYOrPackedQ = Union{QRCompactWYQ{T,S}, QRPackedQ{T,S}} where {T, S}
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/qr.jl]:
```suggestion
QRCompactWYOrPackedQ{T,S} = Union{QRCompactWYQ{T,S}, QRPackedQ{T,S}}
```

**Review by andreasnoack — COMMENTED** [review]
Thanks for fixing this. Please also add a test of this method.

**Comment by andreasnoack** [comment]
@jmkl009 Bump. Please add a test of the new method. That is all that remains before we can merge this PR.

**Comment by jmkl009** [comment]
> @jmkl009 Bump. Please add a test of the new method. That is all that remains before we can merge this PR.

Busy with midterms and stuff, so I need some time to do that.

File: stdlib/LinearAlgebra/src/qr.jl [hunk 2]
```diff
@@ -522,6 +522,10 @@ end
 
 ## Multiplication by Q
 ### QB
+QRCompactWYOrPackedQ{T,S} = Union{QRCompactWYQ{T,S}, QRPackedQ{T,S}}
+mul!(C::StridedVecOrMat{T}, AOrAdjA::Union{Adjoint{<:Any,<:QRCompactWYOrPackedQ{T,S}}, QRCompactWYOrPackedQ{T,S}}, B::StridedVecOrMat{T}) where {T, S} =
+    lmul!(AOrAdjA, copy!(C, B))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/qr.jl]: Shouldn't it be `copyto!`?

**Comment by nrontsis** [comment]
Testings aside, we still need more code to be written to cover all cases already covered by `*` i.e.
```julia
function (*)(A::AbstractQ, b::StridedVector)
function (*)(A::AbstractQ, B::StridedMatrix)
function *(adjQ::Adjoint{<:Any,<:AbstractQ}, B::StridedVecOrMat)
function *(Q::AbstractQ, adjB::Adjoint{<:Any,<:StridedVecOrMat})
function *(adjQ::Adjoint{<:Any,<:AbstractQ}, adjB::Adjoint{<:Any,<:StridedVecOrMat})
function (*)(A::StridedMatrix, Q::AbstractQ)
function *(A::StridedMatrix, adjB::Adjoint{<:Any,<:AbstractQ})
function *(adjA::Adjoint{<:Any,<:StridedVecOrMat}, Q::AbstractQ)
function *(adjA::Adjoint{<:Any,<:StridedVecOrMat}, adjQ::Adjoint{<:Any,<:AbstractQ})
```

@jmkl009 I would be very happy if this were to be completed soon. Would your schedule allow for this? Otherwise, I would be up to take it from here.

File: stdlib/LinearAlgebra/src/qr.jl (hunk 2 above)
Comment (jmkl009) [reply stdlib/LinearAlgebra/src/qr.jl]: For the purpose of this function, they work the same

**Comment by jmkl009** [comment]
> Testing aside, we still need more code to cover all cases already covered by `*` i.e.
> 
> ```julia
> function (*)(A::AbstractQ, b::StridedVector)
> function (*)(A::AbstractQ, B::StridedMatrix)
> function *(adjQ::Adjoint{<:Any,<:AbstractQ}, B::StridedVecOrMat)
> function *(Q::AbstractQ, adjB::Adjoint{<:Any,<:StridedVecOrMat})
> function *(adjQ::Adjoint{<:Any,<:AbstractQ}, adjB::Adjoint{<:Any,<:StridedVecOrMat})
> function (*)(A::StridedMatrix, Q::AbstractQ)
> function *(A::StridedMatrix, adjB::Adjoint{<:Any,<:AbstractQ})
> function *(adjA::Adjoint{<:Any,<:StridedVecOrMat}, Q::AbstractQ)
> function *(adjA::Adjoint{<:Any,<:StridedVecOrMat}, adjQ::Adjoint{<:Any,<:AbstractQ})
> ```
> 
> @jmkl009 I would be very happy if this were to be completed soon. Would your schedule allow for this? Otherwise, I would be up to take it from here.

Sure, you can take it from here. Do not want my busy schedule to hinder the progress overall.

File: stdlib/LinearAlgebra/src/qr.jl (hunk 2 above)
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/qr.jl]: Strange. Back then, I couldn't find a `copy` method instance that would work for matrices... Oh, I just noticed this has been introduced in Julia 1.1. Nevermind!

**Comment by nrontsis** [comment]
I think that this can now be closed due to #31581

**Pull request closed without merging**